portal-kernel/src/com/liferay/portal/kernel/portlet/PortletProviderUtil.java

```diff
@@ -148,29 +132,80 @@ protected static boolean isAssetObject(String className) {
 		return false;
 	}
 
-	private static final ServiceTrackerMap<String, AddPortletProvider>
-		_addServiceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
-			SystemBundleUtil.getBundleContext(), AddPortletProvider.class,
-			"model.class.name");
-	private static final ServiceTrackerMap<String, BrowsePortletProvider>
-		_browseServiceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
-			SystemBundleUtil.getBundleContext(), BrowsePortletProvider.class,
-			"model.class.name");
-	private static final ServiceTrackerMap<String, EditPortletProvider>
-		_editServiceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
-			SystemBundleUtil.getBundleContext(), EditPortletProvider.class,
-			"model.class.name");
-	private static final ServiceTrackerMap<String, ManagePortletProvider>
-		_manageServiceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
-			SystemBundleUtil.getBundleContext(), ManagePortletProvider.class,
-			"model.class.name");
-	private static final ServiceTrackerMap<String, PreviewPortletProvider>
-		_previewServiceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
-			SystemBundleUtil.getBundleContext(), PreviewPortletProvider.class,
-			"model.class.name");
-	private static final ServiceTrackerMap<String, ViewPortletProvider>
-		_viewServiceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
-			SystemBundleUtil.getBundleContext(), ViewPortletProvider.class,
-			"model.class.name");
+	private static final Map
+		<PortletProvider.Action, Map<String, PortletProvider>>
+			_portletProviderActionPortletProviderMap =
+				new ConcurrentHashMap<>();
+	private static final ServiceTracker<PortletProvider, PortletProvider>
+		_portletProviderServiceTracker;
+
+	static {
+		BundleContext bundleContext = SystemBundleUtil.getBundleContext();
+
+		_portletProviderServiceTracker = new ServiceTracker<>(
```

Comment:
@rafaprax For this service tracker, let's do this. Only track PortletProvider, when you want to retrieve portlet provider for a specific action: 1 get the portlet provider 2 if provider exists, check its supported actions include the one you asked.